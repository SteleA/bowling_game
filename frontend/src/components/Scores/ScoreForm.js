import React, {Component} from 'react';
import _ from 'lodash';
import {Button} from 'react-bootstrap';
import ScoreInput from './ScoreInput';

class ScoreForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      scores: {},
    };
  }

  onAddScores(e) {
    if (e) e.preventDefault();
    const rolls = _.map(this.props.round.rolls, roll =>
      ({
        ...roll,
        score: this.getNumber(this.state.scores[roll.id]),
      }));

    const round =
      Object.assign({},
        this.props.round,
        {rolls, score: this.calculateScores(this.state.scores)});

    this.props.addScore(round);
    this.setState({scores: {}});
  }

  getNumber(num) {
    const number = _.toNumber(num);
    return _.isNaN(number) ? 0 : number;
  }

  setScore(id, score) {
    const scores = Object.assign({}, this.state.scores);
    scores[id] = score;
    const totalScore = this.calculateScores(scores) <= 10;
    const lastRound = this.calculateScores(scores) <= 30 && this.props.round.round === 10;
    if ((totalScore || lastRound) && score >= 0) {
      this.setState({scores});
    }
    return false;
  }

  calculateScores(scores) {
    return _.chain(scores)
        .map(score => score)
        .map(_.toNumber)
        .reduce((acc, cur) => acc + cur)
        .value();
  }

  render() {
    return (
      <div>
        {_.map(this.props.round.rolls, roll =>
          <ScoreInput
            id={`score-input-${roll.id}`}
            key={`score-input-${roll.id}`}
            onChange={score => this.setScore(roll.id, score)}
            input={this.getNumber(this.state.scores[roll.id])}
            placeholder={roll.name}
          />
        )}
        <Button block onClick={() => this.onAddScores()}>Submit Scores</Button>
      </div>
    );
  }
}

export default ScoreForm;
