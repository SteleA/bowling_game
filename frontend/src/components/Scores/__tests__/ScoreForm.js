import { shallow, mount } from 'enzyme';  // eslint-disable-line
import { shallowToJson } from 'enzyme-to-json';  // eslint-disable-line
import React from 'react';
import ScoreForm from '../ScoreForm';

const props = {
  round: {
    rolls: [],
  },
};

let mountedComponent;

describe('<ScoreForm />', () => {
  it('should render with default props', () => {
    const wrapper = shallow(
      <ScoreForm {...props} />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  describe('methods', () => {
    beforeEach(() => {
      mountedComponent = mount(<ScoreForm {...props} />);
    });

    describe('getNumber', () => {
      it('return 0', () => {
        expect(mountedComponent.instance().getNumber()).toBe(0);
      });
      it('return 0', () => {
        expect(mountedComponent.instance().getNumber('a')).toBe(0);
      });
      it('return 10', () => {
        expect(mountedComponent.instance().getNumber(10)).toBe(10);
      });
    });

    describe('setScore', () => {
      it('set score', () => {
        mountedComponent.instance().setScore('first', 5);
        expect(mountedComponent.state().scores.first).toBe(5);
      });
      it('score not set score if over 10', () => {
        mountedComponent.instance().setScore('first', 5);
        expect(mountedComponent.state().scores.first).toBe(5);
        mountedComponent.instance().setScore('first', 11);
        expect(mountedComponent.state().scores.first).toBe(5);
      });
      it('set first and second score', () => {
        mountedComponent.instance().setScore('first', 5);
        mountedComponent.instance().setScore('second', 5);
        expect(mountedComponent.state().scores.first).toBe(5);
        expect(mountedComponent.state().scores.second).toBe(5);
      });
      it('not set second score if total score is over 10', () => {
        mountedComponent.instance().setScore('first', 5);
        mountedComponent.instance().setScore('second', 6);
        expect(mountedComponent.state().scores.first).toBe(5);
        expect(mountedComponent.state().scores.second).toBe(undefined);
      });
    });
  });
});
