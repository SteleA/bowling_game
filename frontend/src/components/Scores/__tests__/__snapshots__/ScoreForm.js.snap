// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScoreForm /> should render with default props 1`] = `
<div>
  <Button
    active={false}
    block={true}
    bsClass="btn"
    bsStyle="default"
    disabled={false}
    onClick={[Function]}
  >
    Submit Scores
  </Button>
</div>
`;
